### PR TITLE
Make scopes generic to component type

### PR DESCRIPTION
**Pull request by ocrickard**
This still needs some cleanup, but the tests are passing, and this demonstrates the approach.

We have 2 base classes for components, the standard CKComponent type, then another type specifically for data fetching. We'd like to be able to take the implementation of scopes, and reuse it there so things like component actions and triggers work seamlessly.

I implemented this in 3 different ways, and this way was my favorite. For reference here's also what I tried:

1. Obj-C Generics + select templates. This was a real pain because these two generic systems don't work together.
2. After the above failed, I tried just straight using C++ templates, but the code for this was ugly in my opinion, and I was fairly certain would result in binary bloat.
3. This approach, which uses protocols for components, and controllers, for the methods that are needed.

In all of the above 3 implementations I've had to make the component appearance/bounds animation events generic. I've introduced the idea of a "predicate" that is a normal C function pointer that is passed to a scope root on its initialization. All components and controllers will register with the scope root on construction. After this process, we record every component and controller in the tree that match the predicate. Later, when we want to announce something, we give the scope root the predicate we're interested in, and it allows us to rapidly enumerate over all components or controllers which matched that predicate.

CKComponent and CKComponentController both declare conformance with the protocol.

TODO:

1. Needs tests for the generic predicate system. It's tested through the appearance events and bounds animations right now, but probably some dedicated tests here would help exercise the corner cases of the API.
2. Possibly extend the tests of the bounds animation system. There appear to be minimal tests there.
3. Clean up, naming, copyright notices, etc.

### CODE REVIEW

File: ComponentKit/Core/CKBuildComponent.mm [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+     CKComponent *oldComponent = (CKComponent *)component;
+     id scopeFrameToken = [oldComponent scopeFrameToken];
+     if (scopeFrameToken) {
+       [scopeFrameTokenToOldComponent setObject:oldComponent forKey:scopeFrameToken];
+     }
+   }];
+  
+  __block CKComponentBoundsAnimation boundsAnimation {};
+  [newRoot
+   enumerateComponentsMatchingPredicate:&CKComponentBoundsAnimationPredicate
```
Comment (ocrickard) [inline ComponentKit/Core/CKBuildComponent.mm]: Here's an example of the enumeration system I mentioned.

Comment (ocrickard) [inline ComponentKit/Core/CKBuildComponent.mm]: Before, all of this was implemented inside the scope root itself. Now it's implemented in a few functions above the scope root using a generic API.

File: ComponentKit/Core/Scope/CKComponentControllerAppearanceEvents.h [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+#import <Foundation/Foundation.h>
+
+#import <ComponentKit/CKComponentScopeRoot.h>
+
+BOOL CKComponentControllerAppearanceEventPredicate(id<CKScopedComponentController> controller);
+BOOL CKComponentControllerDisappearanceEventPredicate(id<CKScopedComponentController> controller);
+
+void CKComponentControllerAnnounceAppearance(CKComponentScopeRoot *scopeRoot);
+void CKComponentControllerAnnounceDisappearance(CKComponentScopeRoot *scopeRoot);
```
Comment (ocrickard) [inline ComponentKit/Core/Scope/CKComponentControllerAppearanceEvents.h]: These should rarely, if ever be used directly. Generally they're configured for you when you're using a scope root.

File: ComponentKit/Core/Scope/CKComponentScopeRootFactory.h [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+ *
+ */
+
+#import <Foundation/Foundation.h>
+
+@protocol CKComponentStateListener;
+
+@class CKComponentScopeRoot;
+
+CKComponentScopeRoot *CKComponentScopeRootWithListener(id<CKComponentStateListener> listener);
```
Comment (ocrickard) [inline ComponentKit/Core/Scope/CKComponentScopeRootFactory.h]: This is the normal factory that should be used with CKComponent implementations. For working ranges implementation we'd like to override the predicates for controllers to make the tree traversal more efficient, so may need a way to inject this somehow into the hosting view and datasource. Thoughts?

File: ComponentKit/Core/Scope/CKComponentScopeRootFactory.mm [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+{
+  return [CKComponentScopeRoot
+   rootWithListener:listener
+   componentPredicates:{
+     &CKComponentBoundsAnimationPredicate
+   }
+   componentControllerPredicates:{
+     &CKComponentControllerAppearanceEventPredicate,
+     &CKComponentControllerDisappearanceEventPredicate
+   }];
```
Comment (ocrickard) [inline ComponentKit/Core/Scope/CKComponentScopeRootFactory.mm]: The predicates are just C function pointers.

File: ComponentKit/Core/Scope/CKScopedComponent.h [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+//
+//  Created by Oliver Rickard on 3/27/17.
+//
+//
+
+#import <Foundation/Foundation.h>
+
+@protocol CKScopedComponent <NSObject>
+
+@end
```
Comment (ocrickard) [inline ComponentKit/Core/Scope/CKScopedComponent.h]: I was surprised that the scope infra actually doesn't care about the component type at all, thus this protocol is empty. I still want to keep it so we get type annotations on the id types for which are the components, and which are the controllers in the interfaces of the scope infra. Straight id's are unreadable.

File: ComponentKit/Core/Scope/CKScopedComponentController.h [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+//
+//
+
+#import <Foundation/Foundation.h>
+
+@protocol CKScopedComponent;
+
+@protocol CKScopedComponentController <NSObject>
+
+- (instancetype)initWithComponent:(id<CKScopedComponent>)component;
```
Comment (ocrickard) [inline ComponentKit/Core/Scope/CKScopedComponentController.h]: The scope infra actually constructs component controllers at the moment. I considered pulling this out and passing in a component controller factory, but decided against for now. May need that for section components though.

**Comment by ocrickard** [comment]
@facebook-github-bot import

File: ComponentKit/Core/CKBuildComponent.h [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+struct CKBuildComponentResult {
+  CKComponent *component;
+  CKComponentScopeRoot *scopeRoot;
+  CKComponentBoundsAnimation boundsAnimation;
+};
+
+CKBuildComponentResult CKBuildComponent(CKComponentScopeRoot *previousRoot,
+                                        const CKComponentStateUpdateMap &stateUpdates,
+                                        CKComponent *(^function)(void));
```
Comment (eczarny) [inline ComponentKit/Core/CKBuildComponent.h]: I'm a big fan of moving this function to its own header and source file.

File: ComponentKit/Core/Scope/CKComponentScopeHandle.mm [hunk 8]
```diff
@@ -38,9 +38,7 @@ + (CKComponentScopeHandle *)handleForComponent:(CKComponent *)component
 
   CKComponentScopeHandle *handle = currentScope->stack.top().frame.handle;
   if ([handle acquireFromComponent:component]) {
-    if (CKSubclassOverridesSelector([CKComponent class], [component class], @selector(boundsAnimationFromPreviousComponent:))) {
-      [currentScope->newScopeRoot registerBoundsAnimationComponent:component];
-    }
+    [currentScope->newScopeRoot registerComponent:component];
```
Comment (eczarny) [inline ComponentKit/Core/Scope/CKComponentScopeHandle.mm]: Cool, so register components with the scope root regardless of what it's actually being used for.

File: ComponentKit/Core/Scope/CKComponentScopeHandle.h [hunk 9]
```diff
@@ -26,7 +27,7 @@
  handle corresponding to the current scope. Otherwise it returns nil.
  This is only meant to be called when constructing a component and as part of the implementation itself.
  */
-+ (instancetype)handleForComponent:(CKComponent *)component;
++ (instancetype)handleForComponent:(id<CKScopedComponent>)component;
```
Comment (eczarny) [inline ComponentKit/Core/Scope/CKComponentScopeHandle.h]: Github won't let me add a comment in the implementation file but shouldn't we update the signature in the implementation file as well?

File: ComponentKit/Core/Scope/CKComponentScopeRoot.h [hunk 10]
```diff
@@ -10,18 +10,23 @@
 
 #import <Foundation/Foundation.h>
 
+#import <vector>
+
 #import <ComponentKit/CKComponentBoundsAnimation.h>
```
Comment (eczarny) [inline ComponentKit/Core/Scope/CKComponentScopeRoot.h]: Is this still needed?

File: ComponentKit/Core/Scope/CKComponentScopeRoot.mm [hunk 11]
```diff
@@ -9,126 +9,109 @@
...
 #import "CKComponentScopeRoot.h"
-#import "CKComponentScopeRootInternal.h"
 
 #import <libkern/OSAtomic.h>
 
-#import "CKComponentController.h"
-#import "CKComponentInternal.h"
+#import "CKScopedComponent.h"
+#import "CKScopedComponentController.h"
 #import "CKComponentSubclass.h"
```
Comment (eczarny) [inline ComponentKit/Core/Scope/CKComponentScopeRoot.mm]: Another import that can be removed?

File: ComponentKit/Core/Scope/CKScopedComponentController.h (hunk 6 above)
Comment (eczarny) [inline ComponentKit/Core/Scope/CKScopedComponentController.h]: I think you made the right call. That change alone is fairly substantial, it would make reviewing these changes a bit more difficult.

File: ComponentKit/Core/Scope/CKComponentScopeRootFactory.h (hunk 3 above)
Comment (eczarny) [inline ComponentKit/Core/Scope/CKComponentScopeRootFactory.h]: Hm. I'm not sure we would necessarily want to override the predicates but instead rely on a separate set of predicates. Right? There shouldn't be a need to replace the appearance event predicates when incorporating additional working range predicates.

I'm having a bit of trouble seeing how we could leverage this new scope API outside where it's being used internally by ComponentKit. Let's sync up on this when you have some time.

File: ComponentKit/Core/Scope/CKComponentScopeRootFactory.mm (hunk 4 above)
Comment (eczarny) [inline ComponentKit/Core/Scope/CKComponentScopeRootFactory.mm]: Makes sense, though this does make it a bit harder to implement more efficient traversals. It may be necessary to also pass along some state that can be used for caching the predicate results. What do you think?

File: ComponentKit/Core/Scope/CKComponentScopeHandle.mm (hunk 8 above)
Comment (ocrickard) [inline ComponentKit/Core/Scope/CKComponentScopeHandle.mm]: yep

File: ComponentKit/Core/Scope/CKComponentScopeRootFactory.h (hunk 3 above)
Comment (ocrickard) [inline ComponentKit/Core/Scope/CKComponentScopeRootFactory.h]: Yeah, the predicates we'd want to add would be used in addition to the current predicates for appearance and bounds animations.

File: ComponentKit/Core/Scope/CKComponentScopeRootFactory.mm (hunk 4 above)
Comment (ocrickard) [inline ComponentKit/Core/Scope/CKComponentScopeRootFactory.mm]: Oh to be clear, this is the efficient traversal. We build up the list of matching components on component generation, so by nature this is cached. It happens once for a scope root, and then the results are used over the lifetime.

File: ComponentKit/Core/CKBuildComponent.h [hunk 12]
```diff
@@ -0,0 +1,27 @@
+//
+//  CKBuildComponent.h
+//  ComponentKit
+//
+//  Created by Oliver Rickard on 3/28/17.
```
Comment (adamjernst) [inline ComponentKit/Core/CKBuildComponent.h]: fix header

File: ComponentKit/Core/CKBuildComponent.h [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+//
+//
+
+#import <Foundation/Foundation.h>
+
+#import <ComponentKit/CKComponentScopeTypes.h>
+#import <ComponentKit/CKComponentBoundsAnimation.h>
+
+@protocol CKScopedComponent;
```
Comment (adamjernst) [inline ComponentKit/Core/CKBuildComponent.h]: remove; not used in header

File: ComponentKit/Core/CKBuildComponent.mm (hunk 1 above)
Comment (adamjernst) [inline ComponentKit/Core/CKBuildComponent.mm]: so so good.

File: ComponentKit/Core/CKBuildComponent.mm [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+  NSMapTable *scopeFrameTokenToOldComponent = [NSMapTable strongToStrongObjectsMapTable];
+  [previousRoot
+   enumerateComponentsMatchingPredicate:&CKComponentBoundsAnimationPredicate
+   block:^(id<CKScopedComponent> component) {
+     CKComponent *oldComponent = (CKComponent *)component;
+     id scopeFrameToken = [oldComponent scopeFrameToken];
+     if (scopeFrameToken) {
+       [scopeFrameTokenToOldComponent setObject:oldComponent forKey:scopeFrameToken];
+     }
+   }];
```
Comment (adamjernst) [inline ComponentKit/Core/CKBuildComponent.mm]: you might want to use a std::unordered_map. if you wanted you could even use __unsafe_unretained for the keys since they are guaranteed not to go out of scope during this function's execution.

File: ComponentKit/Core/CKBuildComponent.mm [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+  return boundsAnimation;
+}
+
+CKBuildComponentResult CKBuildComponent(CKComponentScopeRoot *previousRoot,
+                                        const CKComponentStateUpdateMap &stateUpdates,
+                                        CKComponent *(^function)(void))
+{
+  CKThreadLocalComponentScope threadScope(previousRoot, stateUpdates);
+  // Order of operations matters, so first store into locals and then return a struct.
+  CKComponent *component = function();
```
Comment (adamjernst) [inline ComponentKit/Core/CKBuildComponent.mm]: const

File: ComponentKit/Core/CKComponentBoundsAnimationPredicates.h [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+//
+//  Created by Oliver Rickard on 3/28/17.
+//
+//
+
+#import <Foundation/Foundation.h>
+
+#import <ComponentKit/CKComponentScopeRoot.h>
+
+BOOL CKComponentBoundsAnimationPredicate(id<CKScopedComponent> controller);
```
Comment (adamjernst) [inline ComponentKit/Core/CKComponentBoundsAnimationPredicates.h]: docs?

File: ComponentKit/Core/CKBuildComponent.mm (hunk 14 above)
Comment (adamjernst) [inline ComponentKit/Core/CKBuildComponent.mm]: eh actually don't do this

File: ComponentKit/Core/CKComponentControllerInternal.h [hunk 17]
```diff
@@ -11,8 +11,9 @@
 #import <UIKit/UIKit.h>
 
 #import <ComponentKit/CKComponentController.h>
+#import <ComponentKit/CKScopedComponentController.h>
 
-@interface CKComponentController ()
+@interface CKComponentController () <CKScopedComponentController>
```
Comment (adamjernst) [inline ComponentKit/Core/CKComponentControllerInternal.h]: seems kinda mysterious to me, not knowing what it is

File: ComponentKit/Core/Scope/CKComponentControllerAppearanceEvents.h (hunk 2 above)
Comment (adamjernst) [inline ComponentKit/Core/Scope/CKComponentControllerAppearanceEvents.h]: can we just hide these inside the file that uses them instead of making them public?

File: ComponentKit/Core/Scope/CKComponentScopeHandle.h (hunk 9 above)
Comment (adamjernst) [inline ComponentKit/Core/Scope/CKComponentScopeHandle.h]: ahh so this is why the new protocol exists. we can reuse it for… other similar systems.

should we call this `CKScopedObject` instead? `CKScopedElement`? Some other generic name?

File: ComponentKit/Core/Scope/CKComponentControllerAppearanceEvents.h (hunk 2 above)
Comment (ocrickard) [inline ComponentKit/Core/Scope/CKComponentControllerAppearanceEvents.h]: Sadly we have several places that use these.

File: ComponentKit/Core/CKComponentBoundsAnimationPredicates.h (hunk 16 above)
Comment (dshahidehpour) [inline ComponentKit/Core/CKComponentBoundsAnimationPredicates.h]: Not to woodshed too much, but, can we reword this function's name?

`CKComponentHasCustomImplementationForBoundsAnimation()`

or something?

small nit: component, not controller.

File: ComponentKit/Core/Scope/CKComponentScopeRoot.mm [hunk 18]
```diff
@@ -9,126 +9,109 @@
...
-  for (const auto &announceableEvent : announceableEvents()) {
-    if (CKSubclassOverridesSelector([CKComponentController class], [controller class], announceableEvent.second)) {
-      _eventRegistration.insert({{announceableEvent.first, controller}});
+  if (!component) {
+    // Handle this gracefully so we don't have a bunch of nils being passed to predicates.
+    return;
+  }
+  for (const auto &predicate : _componentPredicates) {
+    if (predicate(component)) {
+      _registeredComponents.insert({predicate, component});
```
Comment (dshahidehpour) [inline ComponentKit/Core/Scope/CKComponentScopeRoot.mm]: How does this work? Since the same predicate serves as the same key for multiple components, won't the insertion fail?

File: ComponentKit/Core/Scope/CKComponentControllerAppearanceEvents.h (hunk 2 above)
Comment (dshahidehpour) [inline ComponentKit/Core/Scope/CKComponentControllerAppearanceEvents.h]: I find this method a little confusing because you are providing a scope root. Could we name it something like:

`CKScopeRootAnnounceApperanceToControllers(CKComponentScopeRoot *scopeRoot)`

File: ComponentKit/Core/CKBuildComponent.mm (hunk 14 above)
Comment (ocrickard) [inline ComponentKit/Core/CKBuildComponent.mm]: Yeah, this is just the old code.

File: ComponentKit/Core/CKBuildComponent.mm [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+       CKComponent *oldComponent = [scopeFrameTokenToOldComponent objectForKey:scopeFrameToken];
+       if (oldComponent) {
+         const CKComponentBoundsAnimation ba = [newComponent boundsAnimationFromPreviousComponent:oldComponent];
+         if (ba.duration != 0) {
+           boundsAnimation = ba;
+         }
+       }
+     }
+   }];
+  return boundsAnimation;
```
Comment (ocrickard) [inline ComponentKit/Core/CKBuildComponent.mm]: lol yep! That's the contract with bounds animations. Only a single one gets executed.

File: ComponentKit/Core/Scope/CKComponentScopeRoot.h (hunk 10 above)
Comment (ocrickard) [inline ComponentKit/Core/Scope/CKComponentScopeRoot.h]: Nope, good eye.

File: ComponentKit/Core/Scope/CKComponentScopeRoot.mm (hunk 11 above)
Comment (ocrickard) [inline ComponentKit/Core/Scope/CKComponentScopeRoot.mm]: Probably

File: ComponentKit/Core/Scope/CKComponentScopeRoot.mm (hunk 18 above)
Comment (ocrickard) [inline ComponentKit/Core/Scope/CKComponentScopeRoot.mm]: lol yeah, this was supposed to be a multimap, not a straight unordered_map. Good find.

File: ComponentKit/Core/CKBuildComponent.mm (hunk 19 above)
Comment (dshahidehpour) [inline ComponentKit/Core/CKBuildComponent.mm]: wowow

File: ComponentKit/Core/Scope/CKComponentControllerAppearanceEvents.h (hunk 2 above)
Comment (ocrickard) [inline ComponentKit/Core/Scope/CKComponentControllerAppearanceEvents.h]: I think you're right.

File: ComponentKit/Core/CKBuildComponent.h (hunk 13 above)
Comment (dshahidehpour) [inline ComponentKit/Core/CKBuildComponent.h]: I think we can still remove this

File: ComponentKit/Core/CKBuildComponent.mm [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+#import "CKComponentSubclass.h"
+#import "CKThreadLocalComponentScope.h"
+
+static CKComponentBoundsAnimation boundsAnimationFromPreviousScopeRoot(CKComponentScopeRoot *newRoot, CKComponentScopeRoot *previousRoot)
+{
+  NSMapTable *const scopeFrameTokenToOldComponent = [NSMapTable strongToStrongObjectsMapTable];
+  [previousRoot
+   enumerateComponentsMatchingPredicate:&CKComponentBoundsAnimationPredicate
+   block:^(id<CKScopedComponent> component) {
+     CKComponent *oldComponent = (CKComponent *)component;
```
Comment (dshahidehpour) [inline ComponentKit/Core/CKBuildComponent.mm]: The `CKScopedComponent` protocol is currently empty, maybe we should consider adding `scopeFrameToken` as a required property, then we can avoid the cleanup of this cast in the future?

File: ComponentKit/Core/CKBuildComponent.mm [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+  __block CKComponentBoundsAnimation boundsAnimation {};
+  [newRoot
+   enumerateComponentsMatchingPredicate:&CKComponentBoundsAnimationPredicate
+   block:^(id<CKScopedComponent> component) {
+     CKComponent *newComponent = (CKComponent *)component;
+     id scopeFrameToken = [newComponent scopeFrameToken];
+     if (scopeFrameToken) {
+       CKComponent *oldComponent = [scopeFrameTokenToOldComponent objectForKey:scopeFrameToken];
+       if (oldComponent) {
+         const CKComponentBoundsAnimation ba = [newComponent boundsAnimationFromPreviousComponent:oldComponent];
```
Comment (dshahidehpour) [inline ComponentKit/Core/CKBuildComponent.mm]: maybe it would be worth adding to the `CKScopedComponent`, future implementations can just return a empty implementation.

Do you think that's too much stuff to add to the protocol?

File: ComponentKit/Core/CKBuildComponent.h [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+
+struct CKBuildComponentResult {
+  CKComponent *component;
+  CKComponentScopeRoot *scopeRoot;
+  CKComponentBoundsAnimation boundsAnimation;
+};
+
+CKBuildComponentResult CKBuildComponent(CKComponentScopeRoot *previousRoot,
+                                        const CKComponentStateUpdateMap &stateUpdates,
+                                        CKComponent *(^function)(void));
```
Comment (dshahidehpour) [inline ComponentKit/Core/CKBuildComponent.h]: I know you are just moving code around, but can we give this a better name? `creationBlock`, maybe?

File: ComponentKit/Core/CKComponentBoundsAnimationPredicates.h (hunk 16 above)
Comment (dshahidehpour) [inline ComponentKit/Core/CKComponentBoundsAnimationPredicates.h]: +1

File: ComponentKit/Core/Scope/CKComponentControllerAppearanceEvents.h (hunk 2 above)
Comment (dshahidehpour) [inline ComponentKit/Core/Scope/CKComponentControllerAppearanceEvents.h]: +1 to this still

File: ComponentKit/Core/Scope/CKComponentScopeRootFactory.h (hunk 3 above)
Comment (dshahidehpour) [inline ComponentKit/Core/Scope/CKComponentScopeRootFactory.h]: Just playing devil's advocate, how does someone know to use this factory function, or the initializer for `CKComponentScopeRoot`? Can we make it more clear that this isn't a "plain ol' CKComponentScopeRoot?"

`CKComponentScopeRootWithDefaultPredicates(id<CKComponentStateListner> listener)`?

or, since these predicates are baked into the infra

```
CKComponentScopeRoot::rootWithDefaultPredicates(id<CKComponentStateListener> listener)
```

something like that?

File: ComponentKit/Core/CKBuildComponent.mm (hunk 20 above)
Comment (ocrickard) [inline ComponentKit/Core/CKBuildComponent.mm]: Ahh, interesting! That's a good observation. Let's see if our other implementation has the same problem.

**Comment by ocrickard** [comment]
I smoke-tested this internally, should be ready to merge now. Accept anyone?

**Comment by ocrickard** [comment]
@facebook-github-bot import

**Comment by ocrickard** [comment]
Alright, this landed internally. Merging.
